### PR TITLE
Close temporary files before removal, fix PermissionError

### LINKED ISSUE #123: Permission Error in Windows
Tensoflow cloud version: (0.1.5)
 Python version : 3.7.6 
 OS: Windows 10 64-bit

When running from a notebook:

```
import tensorflow_cloud as tfc
tfc.run(entry_point='mnist.py',
        docker_image_bucket_name="a-test",
)

```
mnist.py file is as in example : 

import tensorflow as tf

```
(x_train, y_train), (_, _) = tf.keras.datasets.mnist.load_data()

x_train = x_train.reshape((60000, 28 * 28))
x_train = x_train.astype('float32') / 255

model = tf.keras.Sequential([
  tf.keras.layers.Dense(512, activation='relu', input_shape=(28 * 28,)),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10, activation='softmax')
])

model.compile(loss='sparse_categorical_crossentropy',
              optimizer=tf.keras.optimizers.Adam(),
              metrics=['accuracy'])

model.fit(x_train,  y_train, epochs=10, batch_size=128)
```

**Result:**

 INFO:tensorflow_cloud.core.containerize:Uploading files to GCS.

INFO:tensorflow_cloud.core.containerize:Building and publishing docker image using Google Cloud Build: gcr.io/dydra- 282320/tf_cloud_train:371ef57f_1db8_4c96_aae6_09e373926732

---------------------------------------------------------------------------
PermissionError                           Traceback (most recent call last)
<ipython-input-11-d0d01e61335b> in <module>
      1 import tensorflow_cloud as tfc
      2 tfc.run(entry_point='mnist.py',
----> 3         docker_image_bucket_name="a-test",
      4 )

~\anaconda3\lib\site-packages\tensorflow_cloud\core\run.py in run(entry_point, requirements_txt, distribution_strategy, docker_base_image, chief_config, worker_config, worker_count, entry_point_args, stream_logs, docker_image_bucket_name, job_labels, **kwargs)
    224     # Delete all the temporary files we created.
    225     if preprocessed_entry_point is not None:
--> 226         os.remove(preprocessed_entry_point)
    227     for f in container_builder.get_generated_files():
    228         os.remove(f)

PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\semih\\AppData\\Local\\Temp\\tmpkjul06z0.py'

### PR DESCRIPTION
**Pull request by christianversloot**
Fixes #123.

TensorFlow Cloud creates temporary files which it then removes out of neatness reasons, but on Windows a `PermissionError` emerges because the files haven't been closed before removal.

This PR fixes the issue by closing the temporary files using their file descriptors before attempting to remove them. In doing so, I attempted to keep the existing definitions backwards compatible by making returning the file descriptors optional.

### CODE REVIEW

**Comment by christianversloot** [comment]
@googlebot I signed it!

**Comment by christianversloot** [comment]
Hi @pavithrasv - it could be impossible to run TF Cloud on Windows machines at the moment (#123), and this PR provides the fix. Would you have some time to look at it?

**Comment by christianversloot** [comment]
@SinaChavoshi, perhaps?

**Comment by pavithrasv** [comment]
This is great, thank you for the PR @christianversloot! I'll take a look at it shortly.

File: src/python/tensorflow_cloud/core/containerize.py [hunk 1]
```diff
@@ -117,15 +119,26 @@ def get_docker_image(
...
         raise NotImplementedError
 
-    def get_generated_files(self):
-        return [self.docker_file_path, self.tar_file_path]
+    def get_generated_files(self, return_descriptors=False):
+        """Get generated file paths and/or descriptors for generated files.
+
+        Args:
+            return_descriptors: Whether to return descriptors as well.
+        """
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/containerize.py:127]: Can you add a returns section to this doc string as well?

File: src/python/tensorflow_cloud/core/run.py [hunk 2]
```diff
@@ -225,9 +226,11 @@ def run(
 
     # Delete all the temporary files we created.
     if preprocessed_entry_point is not None:
+        os.close(pep_file_descriptor)
         os.remove(preprocessed_entry_point)
-    for f in container_builder.get_generated_files():
-        os.remove(f)
+    for file_path, file_descriptor in container_builder.get_generated_files(return_descriptors=True):
```
Comment (pavithrasv) [inline src/python/tensorflow_cloud/core/run.py:231]: Can you format this so that line length <= 80?

**Review by pavithrasv — COMMENTED** [review]
Thank you for the PR!

File: src/python/tensorflow_cloud/core/run.py (hunk 2 above)
Comment (christianversloot) [reply src/python/tensorflow_cloud/core/run.py]: https://github.com/tensorflow/cloud/pull/218/commits/03dd09546460757924104957abb43d5dd1623a89

File: src/python/tensorflow_cloud/core/containerize.py (hunk 1 above)
Comment (christianversloot) [reply src/python/tensorflow_cloud/core/containerize.py:131]: https://github.com/tensorflow/cloud/pull/218/commits/10335ecd4b564384210fe250a7019f946d723008

**Comment by christianversloot** [comment]
Thanks for your review! Commits have been attached.

**Review by pavithrasv — APPROVED** [review]

**Comment by pavithrasv** [comment]
@christianversloot Can you fix some lint errors?

Line too long (96/80) [line-too-long]
	src/python/tensorflow_cloud/core/run.py:197
Trailing whitespace [trailing-whitespace]
	src/python/tensorflow_cloud/core/containerize.py:127
Trailing whitespace [trailing-whitespace]
	src/python/tensorflow_cloud/core/containerize.py:130
Wrong hanging indentation (add 2 spaces).
            (self.docker_file_path, self.docker_file_descriptor),
            ^ |
[bad-continuation]
	src/python/tensorflow_cloud/core/containerize.py:134
Wrong hanging indentation (add 2 spaces).
            (self.tar_file_path, self.tar_file_descriptor)
            ^ |
[bad-continuation]
	src/python/tensorflow_cloud/core/containerize.py:135
Trailing whitespace [trailing-whitespace]
	src/python/tensorflow_cloud/core/preprocess.py:111
Trailing whitespace [trailing-whitespace]
	src/python/tensorflow_cloud/core/preprocess.py:207

**Comment by christianversloot** [comment]
@pavithrasv absolutely!

Running `pylint` locally I think I fixed all the errors with the attached commit. This one, though, was a bit weird to me:

```
Trailing whitespace [trailing-whitespace]
src/python/tensorflow_cloud/core/preprocess.py:111
```

`preprocess.py:111`  - I could not find a trailing whitespace there.
If it persists, please let me know.

**Review by pavithrasv — APPROVED** [review]

**Comment by christianversloot** [comment]
Hi there, any update? :)

**Review by pavithrasv — APPROVED** [review]

**Review by pavithrasv — APPROVED** [review]

**Review by pavithrasv — APPROVED** [review]

**Comment by pavithrasv** [comment]
Fixed a few more lint errors. I will merge this shortly. Thank you and sorry about the wait!

**Pull request merged**